src/lib.rs

```diff
@@ -156,75 +157,26 @@ pub trait Manager: Send + Sync + 'static {
 struct SharedPool<M: Manager> {
     config: ShareConfig,
     manager: M,
-    internals: Mutex<PoolInternals<M::Connection, M::Error>>,
+    internals: Mutex<PoolInternals<M::Connection>>,
     state: PoolState,
-    semaphore: Semaphore,
-}
-
-struct Conn<C, E> {
-    raw: Option<C>,
-    #[allow(dead_code)]
-    last_err: Mutex<Option<E>>,
-    created_at: Instant,
-    last_used_at: Instant,
-    last_checked_at: Instant,
-    brand_new: bool,
+    semaphore: Arc<Semaphore>,
 }
 
-impl<C, E> Conn<C, E> {
-    fn close(&self, state: &PoolState) {
-        state.num_open.fetch_sub(1, Ordering::Relaxed);
-        state.max_idle_closed.fetch_add(1, Ordering::Relaxed);
-        gauge!(OPEN_CONNECTIONS).decrement(1.0);
-        counter!(CLOSED_TOTAL).increment(1);
-    }
-
-    fn expired(&self, timeout: Option<Duration>) -> bool {
-        timeout
-            .and_then(|check_interval| {
-                Instant::now()
-                    .checked_duration_since(self.created_at)
-                    .map(|dur_since| dur_since >= check_interval)
-            })
-            .unwrap_or(false)
-    }
-
-    fn idle_expired(&self, timeout: Option<Duration>) -> bool {
-        timeout
-            .and_then(|check_interval| {
-                Instant::now()
-                    .checked_duration_since(self.last_used_at)
-                    .map(|dur_since| dur_since >= check_interval)
-            })
-            .unwrap_or(false)
-    }
-
-    fn needs_health_check(&self, timeout: Option<Duration>) -> bool {
-        timeout
-            .and_then(|check_interval| {
-                Instant::now()
-                    .checked_duration_since(self.last_checked_at)
-                    .map(|dur_since| dur_since >= check_interval)
-            })
-            .unwrap_or(true)
-    }
-}
-
-struct PoolInternals<C, E> {
+struct PoolInternals<C> {
     config: InternalConfig,
-    free_conns: Vec<Conn<C, E>>,
+    free_conns: Vec<IdleConn<C>>,
     wait_duration: Duration,
     cleaner_ch: Option<Sender<()>>,
 }
 
 struct PoolState {
-    num_open: AtomicU64,
+    num_open: Arc<AtomicU64>,
     max_lifetime_closed: AtomicU64,
-    max_idle_closed: AtomicU64,
+    max_idle_closed: Arc<AtomicU64>,
```

Comment:
Why are atomics wrapped in an `Arc`?
`Arc<T>` is `Send + Safe` as long as `T: Send + Safe`, and `AtomicU64` is already `Send + Safe`.
What do we need the `Arc` for?

Reply:
I think I was wrong.

From https://doc.rust-lang.org/std/sync/atomic/:
> Atomic variables are safe to share between threads (they implement [Sync](https://doc.rust-lang.org/std/marker/trait.Sync.html)) but they do not themselves provide the mechanism for sharing and follow the [threading model](https://doc.rust-lang.org/std/thread/index.html#the-threading-model) of Rust. The most common way to share an atomic variable is to put it into an [Arc](https://doc.rust-lang.org/std/sync/struct.Arc.html) (an atomically-reference-counted shared pointer).

Reply:
Yep, exactly. Those two are needed to be shared becuase `drop` for `ConnState` needs to increment one and decrement another

Reply:
@jkomyno atomics are safely _shareable_ (and you can even safely concurrently mutate them through a shared reference) but they are not magically _shared_ just by cloning them (that would violate the idea of ownership), you need some kind of a reference or a pointer to be able to access the same location in memory from multiple places. If you don't want an `Arc` here, the only other option is for `ConnState` to borrow from the `Pool` and that is not necessarily practical.